File: ios/Classes/Events/NotificareEventStream.swift
//
//  NotificareEventStream.swift
//  notificare
//
//  Created by Helder Pinhal on 03/12/2020.
//

import Foundation

class NotificareEventStream: NSObject, FlutterStreamHandler {

    let eventType: NotificareEventType
    let name: String

    private var eventSink: FlutterEventSink?
    private var pendingEvents: [NotificareEvent] = []

    init(eventType: NotificareEventType) {
        self.eventType = eventType
        self.name = "re.notifica.flutter/events/\(eventType.rawValue)"
    }

    func send(_ event: NotificareEvent) {
        if let sink = self.eventSink {
            sink(event.payload)
        } else {
            pendingEvents.append(event)
        }
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events

        if (self.eventSink != nil) {
            self.pendingEvents.forEach { send($0) }
            self.pendingEvents.removeAll()
        }

        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: notificare_inbox/android/src/main/kotlin/re/notifica/inbox/flutter/events/NotificareEventStream.kt
package re.notifica.inbox.flutter.events

import io.flutter.plugin.common.EventChannel

internal class NotificareEventStream(type: NotificareEvent.Type) : EventChannel.StreamHandler {

    private var eventSink: EventChannel.EventSink? = null
    private val pendingEvents = mutableListOf<NotificareEvent>()

    val name = "re.notifica.inbox.flutter/events/${type.id}"

    fun emit(event: NotificareEvent) {
        val eventSink = this.eventSink

        if (eventSink == null) {
            pendingEvents.add(event)
        } else {
            eventSink.success(event.payload)
        }
    }

    override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink?) {
        this.eventSink = eventSink

        if (eventSink != null) {
            pendingEvents.forEach(::emit)
            pendingEvents.clear()
        }
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
    }
}
